language/move-prover/bytecode/src/read_write_set_analysis.rs

```diff
@@ -556,18 +572,59 @@ impl<'a> TransferFunctions for ReadWriteSetAnalysis<'a> {
                         state.record_access(args[1], Access::Read, func_env)
                     }
                 }
-                Pack(_mid, _sid, _types) => {
+                Pack(_mid, sid, _types) => {
                     // rets[0] = Pack<mid::sid<types>>(args)
+                    for (arg_index, fld) in func_env
+                        .module_env
+                        .get_struct(*sid)
+                        .get_fields()
+                        .enumerate()
+                    {
+                        let ty = fld.get_type();
+                        if ty.is_address() || ty.is_struct() {
+                            if let Some(rhs) = state
+                                .locals
+                                .get_local_node(args[arg_index], func_env)
+                                .cloned()
+                            {
+                                // rets[0]/fld = args[arg_index]
+                                let mut ap = AccessPath::from_index(rets[0], func_env);
+                                ap.add_offset(Offset::field(fld.get_offset()));
+                                // TODO: join, or can we do a strong update?
+                                state.locals.join_access_path(ap, rhs);
+                            }
+                        }
+                    }
                 }
-                Unpack(..) => {
+                Unpack(_mid, sid, _types) => {
                     // rets = Unpack<mid::sid<types>>(args[0])
-                    // pack and unpack touch non-reference values; nothing to do
+                    if state.locals.local_exists(args[0], func_env) {
+                        for (ret_index, fld) in func_env
+                            .module_env
+                            .get_struct(*sid)
+                            .get_fields()
+                            .enumerate()
+                        {
+                            let ty = fld.get_type();
+                            if ty.is_address() || ty.is_struct() {
```

Comment:
Why are we limiting this to addresses and structs?

In a case such as

```
struct S has drop { i: u64 }
fun read_unpacked_u64(s: S): u64 {
        let S { i } = s;
        i
    }
```

I would assume that we want `ret(0) -> Formal(0)/i`, which is the result we get from

```
fun read_u64(s: S): u64 {
        s.i
    }
```

but because of the condition on the types, there will not be any entry for `ret(0)`.
Is this the desired behaviour?

Reply:
Good question. This is an optimization to minimize abstract state and summary size. The analysis seeks to compute an overapproximation of the read/write set, and it only needs to abstract address and reference values to do that. There would be no problem with abstracting other values as well (e.g., we could abstract integers/boolean values and use them to try to prune reads/writes guarded by infeasible path conditions), but the analysis does not currently aspire to this level of precision.

Reply:
Oh, I see, thanks for the clarification.
Just to make sure I understand this properly, what would happen if we had a `Vector<address>` type? As far as I understand, vectors are a native data type, would still `is_struct` return `true`?

For example, with a code such as:

```rust
module 0x2::ReadVector {
    use 0x1::Vector;

    struct S has drop { a: Vector<address> }
    struct Glob has key { b: bool }

    fun extract_addr_from_vec(s: S): address {
        let S { a } = s;
        a[0]
    } // ret |-> { Formal(0)/a/0 }

    fun read_addr_from_vec(s: S): bool acquires Glob {
        let a = extract_addr_from_vec(s);
        *&borrow_global<Glob>(a).b
    }
}
```

(not sure if it's valid move, I could not get this to compile/run because apparently it's not finding stdlib)

I suppose we need to be sure that the read from `Glob` at address `a` is recorded.
Would that work out fine?

Reply:
Yes, this should work as well (I will add it as a test). 

The `is_addr || is_struct` check will indeed return `false` for vectors, but that is ok because this check is only used in the context of pack/unpack instructions, which cannot be performed on a vector

Reply:
> (not sure if it's valid move, I could not get this to compile/run because apparently it's not finding stdlib)

Ah, apologies; this can be resolved by adding `// dep: ../../move-stdlib/modules/Vector.move` at the top of the file. Note that this is an feature of the prover testing framework and doesn't have any meaning elsewhere...

Reply:
Good catch; this did not in fact work! Adding an is_vector check.

Reply:
> Good catch; this did not in fact work! Adding an is_vector check.

I suppose this is only really an issue for `Vector<address>` or `Vector<T>` where `T` directly or indirectly contains an address?

> this can be resolved by adding // dep: ../../move-stdlib/modules/Vector.move at the top of the file

Oh nice, thanks for the tip!

Reply:
> I suppose this is only really an issue for Vector<address> or Vector<T> where T directly or indirectly contains an address?

Correct

Reply:
I did actually add you test locally, but forgot to `git add` it! Will fix in a follow-up.